### PR TITLE
Rewrite journald LogReader and unit test it

### PR DESCRIPTION
**Pull request by corhere**
**- What I did**
I rewrote the journald `LogReader` using a thin cgo wrapper around the `sd_journal` C API and performing all the heavy lifting in idiomatic Go code. I eliminated the need to use journal cursors to check for duplicate log messages by carefully managing the movement of the journal read pointer. The rewritten implementation passes the full `LogReader` test suite and implements the `--follow` flag correctly and performantly. (The claim in #38859 that this is impossible was slightly exaggerated.)

I fixed edge cases in the handling of `--since` with non-monotonic message timestamps in both the journald and logfile readers and extended the `LogReader` suite to regression test it.

**- How I did it**
libsystemd's `sd_journal` API has some known bugs around read-pointer movement. The existing implementation triggered these bugs and worked around them with brute force using cursors. I was able to work around the bugs in the reimplementation so that the read-pointer always moves where it is intended to go, no cursors required.

SystemD provides [a `systemd-journal-remote` command](https://www.freedesktop.org/software/systemd/man/systemd-journal-remote.service.html) which can be used to write arbitrary journal entries to journal files at arbitrary paths. I wrote a fake `journal.Send` which uses that command to write journal entries to isolated, temporary journal files which are independent from the system or user journals. This is what makes it possible to close the loop between the journal log writer, libsystemd and the journal log reader in unit tests.

**- How to verify it**
```console
$ make BIND_DIR=. shell
# go test -tags journald ./daemon/logger/...
```

**- Description for the changelog**
* Newlines are no longer lost from container logs when using the `journald` log driver.
* The non-user-defined attributes `CONTAINER_PARTIAL_ID`, `CONTAINER_PARTIAL_ORDINAL`, `CONTAINER_PARTIAL_MESSAGE`, `CONTAINER_PARTIAL_LAST` and `IMAGE_NAME` are no longer output from `docker logs --details` when using the `journald` log driver.
* Resolved an issue where log messages could be incorrectly filtered out when using `docker logs --since` on containers using the `json-file`, `local` or `journald` log drivers if the system clock had been rolled back to before the `--since` timestamp while the container was running.

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by corhere** [comment]
Note for reviewers: the sequence of commits escalates from minor refactorings up to a complete rewrite which clobbers  large swathes of code touched in prior commits. I have done this to capture (a cleaned up version of) the steps I took which led up to the rewrite in git history and so that we have multiple bisection points to revert to in case I have introduced a regression. In light of this, I recommend starting by reviewing the squashed diff rather than the individual commits.

File: daemon/logger/journald/read.go [hunk 1]
```diff
@@ -235,7 +236,7 @@ func (s *journald) followJournal(logWatcher *logger.LogWatcher, j *C.sd_journal,
 		}
 		select {
 		case <-logWatcher.WatchConsumerGone():
-			break // won't be able to write anything anymore
+			break loop // won't be able to write anything anymore
```
Comment (kolyshkin) [inline daemon/logger/journald/read.go]: Does not look related to the commit message.

File: daemon/logger/journald/read.go [hunk 2]
```diff
@@ -365,11 +358,7 @@ func (s *journald) readLogs(logWatcher *logger.LogWatcher, config logger.ReadCon
 	}
 	if config.Follow {
 		cursor = s.followJournal(logWatcher, j, cursor, untilUnixMicro)
```
Comment (tianon) [inline daemon/logger/journald/read.go]: We don't seem to (currently) `close(logWatcher.Msg)` if we get this far - will this new `defer` have a negative effect on `followJournal`?

**Review by tianon — APPROVED** [review]

File: daemon/logger/journald/read.go (hunk 2 above)
Comment (corhere) [reply daemon/logger/journald/read.go]: I think that's a mistake in my most recent rebase; the `defer close(logWatcher.Msg)` line at the top of `followJournal` was supposed to be dropped in this commit

**Review by cpuguy83 — APPROVED** [review]

File: Dockerfile [hunk 3]
```diff
@@ -386,9 +386,6 @@ RUN --mount=type=cache,sharing=locked,id=moby-dev-aptlib,target=/var/lib/apt \
             dbus-user-session \
             systemd \
             systemd-sysv
-RUN mkdir -p hack \
-  && curl -o hack/dind-systemd https://raw.githubusercontent.com/AkihiroSuda/containerized-systemd/b70bac0daeea120456764248164c21684ade7d0d/docker-entrypoint.sh \
-  && chmod +x hack/dind-systemd
```
Comment (thaJeztah) [inline Dockerfile:391]: ./cc @AkihiroSuda (FYI - this commit adds a copy/variation of the above in `hack/`)

(I recall there originally was some discussion about this in https://github.com/moby/moby/pull/40493#discussion_r383409927)

File: daemon/logger/journald/internal/fake/sender.go [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+// ErrCommandNotFound is returned when the systemd-journal-remote command could
+// not be located at the well-known paths or $PATH.
+var ErrCommandNotFound = errors.New("systemd-journal-remote command not found")
+
+// JournalRemoteCmdPath searches for the systemd-journal-remote command in
+// well-known paths and the directories named in the $PATH environment variable.
+func JournalRemoteCmdPath() (string, error) {
+	for _, p := range cmdPaths {
+		if path, err := exec.LookPath(p); err == nil {
```
Comment (thaJeztah) [inline daemon/logger/journald/internal/fake/sender.go:48]: ~Wondering if (in a follow-up) we should use `golang.org/x/sys/execabs` for this; https://github.com/golang/sys/blob/b6088ccd6cba9272b1e5639ef2df5b5731b85abc/execabs/execabs.go#L5-L13~

edit: Ignore me; this is an internal package and only for testing

File: daemon/logger/journald/internal/fake/sender.go [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+	"github.com/coreos/go-systemd/v22/journal"
+	"gotest.tools/v3/assert"
+
+	"github.com/docker/docker/daemon/logger/journald/internal/export"
+)
+
+// The systemd-journal-remote command is not conventionally installed on $PATH.
+// The manpage from upstream systemd lists the command as
+// /usr/lib/systemd/systemd-journal-remote, but Debian installs it to
+// /lib/systemd instead.
```
Comment (thaJeztah) [inline daemon/logger/journald/internal/fake/sender.go:33]: This is awkward. Well.. it is what it is 🤷‍♂️

**Review by thaJeztah — APPROVED** [review]
LGTM

did not look into all bits fully in-depth, but went through the commits a couple of times; looks great!

File: Dockerfile (hunk 3 above)
Comment (AkihiroSuda) [reply Dockerfile:391]: Copying the script SGTM, but what is the modification here?
